File: mobile/src/main/java/uk/co/appsbystudio/geoshare/utils/ui/notifications/NewShareLocationNotification.java
package uk.co.appsbystudio.geoshare.utils.ui.notifications;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Build;
import android.support.v4.app.NotificationCompat;

import uk.co.appsbystudio.geoshare.MainActivity;
import uk.co.appsbystudio.geoshare.R;

public class NewShareLocationNotification {

    private static final String NOTIFICATION_TAG = "NewShareLocation";

    public static void notify(final Context context,
                              final String exampleString, final int number) {
        final Resources res = context.getResources();


        final String title = res.getString(R.string.new_share_location_notification_title_template, exampleString);
        final String text = res.getString(R.string.new_share_location_notification_placeholder_text_template);

        final NotificationCompat.Builder builder = new NotificationCompat.Builder(context, "share_channel")
                .setDefaults(Notification.DEFAULT_ALL)
                .setSmallIcon(R.drawable.ic_stat_new_share_location)
                .setContentTitle(title)
                .setContentText(text)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setTicker(exampleString)
                .setNumber(number)
                .setContentIntent(
                        PendingIntent.getActivity(
                                context,
                                0,
                                new Intent(context, MainActivity.class),
                                PendingIntent.FLAG_UPDATE_CURRENT))
                .setStyle(new NotificationCompat.BigTextStyle()
                        .bigText(text)
                        .setBigContentTitle(title)
                        .setSummaryText("New location"))
                .setAutoCancel(true);

        notify(context, builder.build());
    }

    @TargetApi(Build.VERSION_CODES.ECLAIR)
    private static void notify(final Context context, final Notification notification) {
        final NotificationManager nm = (NotificationManager) context
                .getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                NotificationChannel notificationChannel = new NotificationChannel("share_channel", "New location notification", NotificationManager.IMPORTANCE_DEFAULT);
                nm.createNotificationChannel(notificationChannel);
            }
            nm.notify(NOTIFICATION_TAG, 0, notification);
        }
    }

    /**
     * Cancels any notifications of this type previously shown using
     * {@link #notify(Context, String, int)}.
     */
    @TargetApi(Build.VERSION_CODES.ECLAIR)
    public static void cancel(final Context context) {
        final NotificationManager nm = (NotificationManager) context
                .getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm != null) {
            nm.cancel(NOTIFICATION_TAG, 0);
        }
    }
}

File: mobile/src/main/java/uk/co/appsbystudio/geoshare/utils/ui/notifications/NewShareLocationNotification.kt
package uk.co.appsbystudio.geoshare.utils.ui.notifications

import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationCompat
import uk.co.appsbystudio.geoshare.R

object NewShareLocationNotification {

    private const val NOTIFICATION_TAG = "NewShareLocation"

    fun notify(context: Context,
               exampleString: String?, number: Int) {
        val res = context.resources


        val title = res.getString(R.string.new_share_location_notification_title_template, exampleString)
        val text = res.getString(R.string.new_share_location_notification_placeholder_text_template)

        val builder = NotificationCompat.Builder(context, "share_channel").apply {
            setDefaults(Notification.DEFAULT_ALL)
            setSmallIcon(R.drawable.ic_stat_new_share_location)
            setContentTitle(title)
            priority = NotificationCompat.PRIORITY_DEFAULT
            setTicker(exampleString)
            setNumber(number)
            setStyle(NotificationCompat.BigTextStyle()
                    .setBigContentTitle(title)
                    .setSummaryText("New location"))
            setAutoCancel(true)
        }

        notify(context, builder.build())
    }

    @TargetApi(Build.VERSION_CODES.ECLAIR)
    private fun notify(context: Context, notification: Notification) {
        val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val notificationChannel = NotificationChannel("share_channel", "New location notification", NotificationManager.IMPORTANCE_DEFAULT)
            nm.createNotificationChannel(notificationChannel)
        }
        nm.notify(NOTIFICATION_TAG, 0, notification)

    }

    @TargetApi(Build.VERSION_CODES.ECLAIR)
    fun cancel(context: Context) {
        val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        nm.cancel(NOTIFICATION_TAG, 0)
    }
}
